File: code/java/exercises/burrowsDelta/burrowsDelta/Bag.java
package burrowsDelta;

/**
 * Bag
 * 
 * Interface to serve as an example of the bag (multi-set) ADT.
 * This differs from what Sedgewick calls a "bag" (but
 * he's wrong). I relied somewhat on the Bag interface
 * in org.apache.commons.collections to define what a bag
 * should be (for example, the iterator returns each item
 * the number of times it's in the bag, not necessarily only
 * once for every unique item).
 * 
 * If overridden, the toString method for bag should return
 * a comma separate list of the elements of the bag bordered
 * by square braces. For instance a bag with elements A,B, and C
 * could return the String "[A, B, C]".
 * @author Thomas VanDrunen
 * Algorithmic Commonplaces
 * July 9, 2014
 * @param <E> The base-type of the bag
 */

public interface Bag<E> extends Iterable<E> {

    /**
     * Add an item to the bag, increasing its count if it's
     * already there.
     * @param item The item to add
     */
    void add(E item);

    /**
     * How many times does this bag contain this item?
     * @param item The item to check
     * @return The number of occurrences of this item in the bag
     */
    int count(E item);

    /**
     * Remove (all occurrences of) an item from the bag, if it's there
     * (ignore otherwise).
     * @param item The item to remove
     */
    void remove(E item);

    /**
     * The number of items in the bag. This is the sum
     * of the counts, not the number of unique items.
     * @return The number of items.
     */
    int size();

    /**
     * Is the bag empty?
     * @return True if the bag is empty, false otherwise.
     */
    boolean isEmpty();



}

File: code/python/huffman/bag.py
'''
Created on Jul 16, 2014

@author: tvandrun
'''



from collections.abc import Collection

from abc import abstractmethod

class Bag(Collection):
    # Inherts __contains__, __iter__, and __len__ from Collection
    
    # Add an item to the bag, increasing its count if it's there already
    @abstractmethod
    def add(self, x):
        pass
    
    # How many times does this bag contain this item?
    # Equivalent to Bag.count() in Java examples.  
    @abstractmethod
    def __getitem__(self, item):
        pass
    
    # Remove (all occurrences of) an item from the bag, if it's there.
    # Equivalent to Bag.remove() in Java examples.
    @abstractmethod
    def discard(self, item):
        pass
    
    


class DictBag(Bag):
    
    def __init__(self):
        self.internal = {}
        
    def add(self, item):
        if item in self.internal :
            self.internal[item] = self.internal[item] + 1
        else :
            self.internal[item] = 1
    
    def __getitem__(self, item):
        if item in self.internal :
            return self.internal[item]
        else:
            return 0;
    
    def discard(self, item):
        if item in self.internal :
            del self.internal[item]
            
    def __len__(self):
        return sum([self.internal[item] for item in self.internal])
    
    def __iter__(self):
        for item in self.internal :
            for i in range(self.internal[item]) :
                yield item

    def __contains__(self, item):
        return self[item] != 0
    
    def __str__(self):
        return "[" + ",".join([str(item) + ":" + str(self[item]) for item in set(self)]) + "]"
                
class ListBag(Bag):
    
    def __init__(self):
        self.internal = []
        
    def add(self, item):
        self.internal.append(item)
    
    def __getitem__(self, item):
        c = 0
        for x in self.internal :
            if item == x :
                c += 1
        return c
    
    def discard(self, item):
        self.internal = [x for x in self.internal if x != item]
        
    def __len__(self):
        return len(self.internal)
    
    def __iter__(self):
        return iter(self.internal)
    
    def __contains__(self, item):
        return self[item] != 0

